src/components/FeaturesDemo/index.tsx

```diff
@@ -0,0 +1,726 @@
+    PAUSE_MD,
+    PAUSE_SM,
+    PAUSE_XS,
+} from "@site/src/components/FeaturesDemo/constants";
+import { WebBrowser, WebBrowserApi } from "@site/src/components/FeaturesDemo/WebBrowser";
+import { PRODUCTS_SAMPLE_DATA } from "@site/src/components/FeaturesDemo/WebBrowser/AppPage/constants";
+import { AppPage } from "@site/src/components/FeaturesDemo/WebBrowser/AppPage";
+import { ReportPage } from "@site/src/components/FeaturesDemo/WebBrowser/ReportPage";
+import {
+    DIFF_ATTEMPT,
+    SUCCESS_ATTEMPT,
+    TESTS_SAMPLE_DATA,
+} from "@site/src/components/FeaturesDemo/WebBrowser/ReportPage/constants";
+import { Status, TestState } from "@site/src/components/FeaturesDemo/WebBrowser/ReportPage/types";
+import {
+    AnimatedCursor,
+    AnimatedCursorApi,
+} from "@site/src/components/FeaturesDemo/AnimatedCursor";
+import {
+    ComponentTypes,
+    PageTypes,
+    ReportPageData,
+    useStore,
+    WebBrowserWindowData,
+} from "@site/src/components/FeaturesDemo/store";
+import {
+    pause,
+    ScreenOrientation,
+    scrollToLineIfNeeded,
+    typeCode,
+    typeConsoleMessage,
+    useScreenOrientation,
+} from "@site/src/components/FeaturesDemo/utils";
+
+export function FeaturesDemo(props: { className?: string }) {
+    const store = useStore();
+    const heading = useStore(state => state.heading);
+
+    const appBrowserPageRef = useRef<WebBrowserApi>(null);
+    const reportBrowserPageRef = useRef<WebBrowserApi>(null);
+    const reportCursorRef = useRef<AnimatedCursorApi>(null);
+
+    const codeRefs = CODE_MESSAGES.map(section => ({
+        ...section,
+        ref: useRef<HTMLSpanElement>(null),
+    }));
+    const codeElementRef = useRef<HTMLDivElement>(null);
+    const codeCleanupRef = useRef<() => void>(null);
+
+    const codeElement = (
+        <div
+            className="relative h-full w-full overflow-y-scroll scroll-smooth p-1"
+            ref={codeElementRef}
+        >
+            <div
+                className="absolute text-right text-blue-400 opacity-40"
+                dangerouslySetInnerHTML={{
+                    __html: Array.from(Array(50).keys()).join("<br>"),
+                }}
+            ></div>
+            <div className="pl-7">
+                {codeRefs.map(({ ref }, index) => (
+                    <span ref={ref} key={index}></span>
+                ))}
+            </div>
+        </div>
+    );
+
+    const consoleRefs = CONSOLE_MESSAGES.map(message => ({
+        ...message,
+        ref: useRef<HTMLSpanElement>(null),
+    }));
+    const consoleCleanupRef: MutableRefObject<(() => void) | null> = useRef<() => void>(null);
+
+    const consoleElement = (
+        <>
+            {consoleRefs.map(({ messages, ref }, index) => (
+                <span ref={ref} key={`${messages[0]}_${index}`}></span>
+            ))}
+        </>
+    );
+
+    const controller = new AbortController();
+    const signal = controller.signal;
+
+    const steps: (() => void | Promise<void>)[] = [
```

Comment:
imho, it would be better to move stept to the separate file(files) for more readability

Reply:
yeah, this thought has crossed my mind, however A LOT of things (many refs, signals, store data, etc.) are used inside steps (via closures), if we move them outside, we'll severely worsen readability in that aspect, so in the end I decided against it.

Otherwise we'll need to construct a function which takes 20+ params (even if packed into an object) and write corresponding types for them all.

Do you think it's worth it?